File: compiler/engine/parser.py
"""
Shannon Model Parser
Parses computational graphs from standard ONNX protobuf binaries and JSON definitions into Shannon IR.
Enforces strict schema validation, truthful data extraction, and rejects unsupported operators without silent fallbacks.
"""

import os
from typing import Dict, Any, List, Optional
import numpy as np
import onnx
from onnx import numpy_helper
from .ir import ModelGraph, Tensor, Layer

# Supported ONNX operators mapped to Shannon canonical operator types
SUPPORTED_ONNX_OPS = {
    "Conv": "Conv2D",
    "Gemm": "Dense",
    "MatMul": "Dense",
    "Relu": "Relu",
    "MaxPool": "MaxPool2D",
    "AveragePool": "AvgPool2D",
    "GlobalAveragePool": "GlobalAvgPool2D",
    "Add": "Add",
    "Flatten": "Flatten",
    "Reshape": "Reshape",
    "BatchNormalization": "BatchNorm",
}

class ModelParser:
    @staticmethod
    def parse_dict(data: Dict[str, Any]) -> ModelGraph:
        """
        Parses a structured dictionary / JSON into Shannon ModelGraph IR.
        Strictly validates schema, rejecting malformed definitions.
        """
        if not isinstance(data, dict):
            raise TypeError(f"Model definition must be a dictionary, got {type(data).__name__}")
            
        if "layers" not in data:
            raise KeyError("Missing required 'layers' list in model definition.")
            
        layers_data = data["layers"]
        if not isinstance(layers_data, list):
            raise TypeError(f"Field 'layers' must be a list, got {type(layers_data).__name__}")
            
        if len(layers_data) == 0:
            raise ValueError("Model must contain at least one layer.")
            
        name = data.get("name", "CustomModel")
        graph = ModelGraph(name)
        
        # 1. Parse Tensors
        tensors_data = data.get("tensors", {})
        if not isinstance(tensors_data, dict):
            raise TypeError(f"Field 'tensors' must be a dictionary, got {type(tensors_data).__name__}")
            
        for t_name, t_info in tensors_data.items():
            if not isinstance(t_info, dict):
                raise TypeError(f"Tensor '{t_name}' specification must be a dictionary.")
            shape = tuple(t_info.get("shape", [1]))
            dtype = t_info.get("dtype", "float32")
            raw_data = t_info.get("data")
            np_data = np.array(raw_data, dtype=np.float32) if raw_data is not None else None
            tensor = Tensor(t_name, shape, dtype, np_data)
            graph.add_tensor(tensor)

        # 2. Parse Layers
        for idx, l_info in enumerate(layers_data):
            if not isinstance(l_info, dict):
                raise TypeError(f"Layer at index {idx} must be a dictionary.")
                
            required_keys = ["layer_id", "op_type", "inputs", "outputs"]
            missing = [k for k in required_keys if k not in l_info]
            if missing:
                raise KeyError(f"Layer at index {idx} is missing required fields: {missing}")
                
            layer = Layer(
                layer_id=l_info["layer_id"],
                op_type=l_info["op_type"],
                inputs=list(l_info["inputs"]),
                outputs=list(l_info["outputs"]),
                params=l_info.get("params", {})
            )
            
            # Attach weights if present
            w_info = l_info.get("weights")
            if w_info and isinstance(w_info, dict):
                w_shape = tuple(w_info.get("shape", [1, 1]))
                w_data = np.array(w_info["data"], dtype=np.float32) if "data" in w_info else np.zeros(w_shape, dtype=np.float32)
                layer.weights = Tensor(f"{layer.layer_id}_w", w_shape, "float32", w_data)
            
            b_info = l_info.get("bias")
            if b_info and isinstance(b_info, dict):
                b_shape = tuple(b_info.get("shape", [1]))
                b_data = np.array(b_info.get("data"), dtype=np.float32) if "data" in b_info else np.zeros(b_shape, dtype=np.float32)
                layer.bias = Tensor(f"{layer.layer_id}_b", b_shape, "float32", b_data)

            graph.add_layer(layer)

        default_inputs = [graph.layers[0].inputs[0]] if (graph.layers and graph.layers[0].inputs) else []
        default_outputs = [graph.layers[-1].outputs[0]] if (graph.layers and graph.layers[-1].outputs) else []
        graph.inputs = data.get("inputs", default_inputs)
        graph.outputs = data.get("outputs", default_outputs)
        graph.compute_stats()
        return graph

    @staticmethod
    def parse_onnx(onnx_path: str) -> ModelGraph:
        """
        Parses a standard ONNX protobuf model file into a genuine Shannon ModelGraph IR.
        Extracts actual initializers, parameters, layer connectivity, and tensor shapes.
        Throws explicit error if an unsupported operator is encountered (no silent fallbacks).
        """
        if not os.path.exists(onnx_path):
            raise FileNotFoundError(f"ONNX model file not found: {onnx_path}")

        try:
            model = onnx.load(onnx_path)
        except Exception as e:
            raise ValueError(f"Malformed or unreadable ONNX file '{os.path.basename(onnx_path)}': {str(e)}")

        graph_def = model.graph
        graph_name = graph_def.name or os.path.splitext(os.path.basename(onnx_path))[0]
        graph = ModelGraph(graph_name)

        # 1. Extract Initializers (Weights, Biases, Constants)
        weights_dict: Dict[str, np.ndarray] = {}
        for init in graph_def.initializer:
            try:
                arr = numpy_helper.to_array(init)
                weights_dict[init.name] = arr
            except Exception as e:
                raise ValueError(f"Failed to decode ONNX initializer '{init.name}': {str(e)}")

        # 2. Extract Graph Value Information (Inputs / Outputs)
        for val_info in list(graph_def.input) + list(graph_def.output) + list(graph_def.value_info):
            t_name = val_info.name
            if t_name in weights_dict:
                continue
            shape_list = []
            if val_info.type.tensor_type.HasField("shape"):
                for d in val_info.type.tensor_type.shape.dim:
                    if d.HasField("dim_value"):
                        shape_list.append(d.dim_value)
                    else:
                        shape_list.append(1)  # Dynamic batch dimension normalized to 1
            shape = tuple(shape_list) if shape_list else (1, 16)
            graph.add_tensor(Tensor(t_name, shape, "float32"))

        # 3. Parse Nodes & Map to Shannon Layers
        for idx, node in enumerate(graph_def.node):
            op_type_onnx = node.op_type
            if op_type_onnx not in SUPPORTED_ONNX_OPS:
                supported_list = ", ".join(sorted(SUPPORTED_ONNX_OPS.keys()))
                raise ValueError(
                    f"Unsupported ONNX operator '{op_type_onnx}' at node '{node.name or idx}'. "
                    f"Shannon TinyML static compiler supports: {supported_list}."
                )

            canonical_op = SUPPORTED_ONNX_OPS[op_type_onnx]
            node_id = node.name or f"{op_type_onnx.lower()}_{idx}"

            # Extract node attributes
            params: Dict[str, Any] = {}
            for attr in node.attribute:
                if attr.type == onnx.AttributeProto.INTS:
                    params[attr.name] = list(attr.ints)
                elif attr.type == onnx.AttributeProto.INT:
                    params[attr.name] = attr.i
                elif attr.type == onnx.AttributeProto.FLOAT:
                    params[attr.name] = attr.f
                elif attr.type == onnx.AttributeProto.STRING:
                    params[attr.name] = attr.s.decode("utf-8")

            # Identify input activations vs weights/biases
            activation_inputs = []
            weight_tensor = None
            bias_tensor = None

            for inp in node.input:
                if inp in weights_dict:
                    arr = weights_dict[inp]
                    if weight_tensor is None and len(arr.shape) >= 2:
                        weight_tensor = Tensor(f"{node_id}_w", tuple(arr.shape), "float32", arr.astype(np.float32))
                    elif bias_tensor is None or len(arr.shape) == 1:
                        bias_tensor = Tensor(f"{node_id}_b", tuple(arr.shape), "float32", arr.astype(np.float32))
                elif inp:
                    activation_inputs.append(inp)

            outputs = list(node.output)

            # Ensure output tensor exists in graph
            for out in outputs:
                if out not in graph.tensors:
                    # Infer output shape from weight or input
                    out_shape = (1, 32)
                    if weight_tensor:
                        out_shape = (1, weight_tensor.shape[0]) if len(weight_tensor.shape) == 2 else (1, weight_tensor.shape[0], 16, 16)
                    graph.add_tensor(Tensor(out, out_shape, "float32"))

            layer = Layer(
                layer_id=node_id,
                op_type=canonical_op,
                inputs=activation_inputs,
                outputs=outputs,
                params=params
            )
            layer.weights = weight_tensor
            layer.bias = bias_tensor
            graph.add_layer(layer)

        # Set graph inputs and outputs
        graph.inputs = [inp.name for inp in graph_def.input if inp.name not in weights_dict]
        graph.outputs = [out.name for out in graph_def.output]
        if not graph.inputs and graph.layers:
            graph.inputs = graph.layers[0].inputs
        if not graph.outputs and graph.layers:
            graph.outputs = graph.layers[-1].outputs

        graph.compute_stats()
        return graph

File: frontend/src/compiler/parser.ts
import { ModelGraph, Tensor, Layer } from './ir';

export class ModelParser {
  /**
   * Parses arbitrary JSON model graphs or layer lists into a canonical Shannon ModelGraph.
   */
  static parseDict(data: Record<string, any>): ModelGraph {
    const name = data.name || data.model_name || 'CustomModel';
    const graph = new ModelGraph(name);

    // 1. Parse Tensors
    const tensorsData = data.tensors || {};
    if (typeof tensorsData === 'object' && !Array.isArray(tensorsData)) {
      for (const [tName, tInfo] of Object.entries<any>(tensorsData)) {
        const shape = Array.isArray(tInfo.shape) ? tInfo.shape : [1];
        const dtype = tInfo.dtype || 'float32';
        const rawData = tInfo.data;
        const tensor = new Tensor(tName, shape, dtype, rawData);
        graph.addTensor(tensor);
      }
    }

    // 2. Parse Layers
    const layersData: any[] = Array.isArray(data.layers) ? data.layers : [];
    if (layersData.length === 0) {
      throw new Error(`Model '${name}' contains no layers. Provide a valid 'layers' array.`);
    }

    for (let idx = 0; idx < layersData.length; idx++) {
      const lInfo = layersData[idx];
      const layerId = lInfo.layer_id || lInfo.name || `layer_${idx}`;
      const opType = lInfo.op_type || lInfo.class_name || lInfo.type || 'Dense';
      const inputs = Array.isArray(lInfo.inputs) ? lInfo.inputs : (lInfo.input ? [lInfo.input] : [`t_in_${idx}`]);
      const outputs = Array.isArray(lInfo.outputs) ? lInfo.outputs : (lInfo.output ? [lInfo.output] : [`t_out_${idx}`]);
      const params = lInfo.params || lInfo.config || {};

      const layer = new Layer(layerId, opType, inputs, outputs, params);

      // Attach weights if specified
      if (lInfo.weights) {
        const wShape = Array.isArray(lInfo.weights.shape) ? lInfo.weights.shape : [1, 1];
        const wData = lInfo.weights.data || null;
        layer.weights = new Tensor(`${layerId}_w`, wShape, 'float32', wData);
      } else if (opType === 'Dense' || opType === 'FullyConnected') {
        const inFeat = params.in_features || params.units_in || 32;
        const outFeat = params.out_features || params.units || 16;
        layer.weights = new Tensor(`${layerId}_w`, [inFeat, outFeat], 'float32');
        layer.bias = new Tensor(`${layerId}_b`, [outFeat], 'float32');
      } else if (opType === 'Conv2D' || opType === 'Conv1D' || opType === 'DepthwiseConv2D') {
        const inCh = params.in_channels || 8;
        const outCh = params.out_channels || 16;
        const kh = params.kernel_h || params.kernel_size || 3;
        const kw = params.kernel_w || 1;
        layer.weights = new Tensor(`${layerId}_w`, [outCh, inCh, kh, kw], 'float32');
        layer.bias = new Tensor(`${layerId}_b`, [outCh], 'float32');
      }

      graph.addLayer(layer);

      // Ensure tensors exist in graph
      for (const inT of inputs) {
        if (!graph.tensors[inT]) {
          graph.addTensor(new Tensor(inT, lInfo.in_shape || [1, 32], 'float32'));
        }
      }
      for (const outT of outputs) {
        if (!graph.tensors[outT]) {
          graph.addTensor(new Tensor(outT, lInfo.out_shape || [1, 16], 'float32'));
        }
      }
    }

    graph.inputs = Array.isArray(data.inputs) ? data.inputs : (graph.layers.length > 0 ? graph.layers[0].inputs : []);
    graph.outputs = Array.isArray(data.outputs) ? data.outputs : (graph.layers.length > 0 ? graph.layers[graph.layers.length - 1].outputs : []);
    graph.computeStats();
    return graph;
  }

  /**
   * Parses an uploaded model file into a Shannon ModelGraph.
   * Strictly enforces explicit errors - never silently substitutes another model based on filename.
   */
  static async parseFile(file: File): Promise<ModelGraph> {
    const fileName = file.name.toLowerCase();

    if (fileName.endsWith('.json')) {
      const text = await file.text();
      try {
        const parsed = JSON.parse(text);
        return ModelParser.parseDict(parsed);
      } catch (err: any) {
        throw new Error(`JSON Syntax Error in ${file.name}: ${err.message}`);
      }
    } else if (fileName.endsWith('.onnx')) {
      // In the browser client without server ONNX protobuf runtime, never silently fall back
      // to another model. Always raise an explicit descriptive error.
      throw new Error(
        `Unsupported ONNX model format in '${file.name}'. Client-side parser requires Shannon IR JSON. Please export your model using the Shannon Python exporter (compiler/api.py) or upload a Shannon JSON graph.`
      );
    } else {
      throw new Error(
        `Unsupported file format '${file.name}'. Please upload a Shannon IR model JSON file (.json).`
      );
    }
  }
}
